tf/tfprocess.py

```diff
@@ -1199,6 +1324,80 @@ def construct_net(self, inputs):
             flow = self.residual_block(flow,
                                        self.RESIDUAL_FILTERS,
                                        name='residual_{}'.format(i + 1))
+        return flow
+
+    def create_encoder_body(self, inputs, embedding_size):
+        # Policy head
+        assert self.POLICY_HEAD == pb.NetworkFormat.POLICY_ATTENTION
+
+        # TODO: re-add support for policy encoder blocks
+        # do some input processing
+        if self.use_smolgen:
+            self.smol_weight_gen_dense = tf.keras.layers.Dense(
+                64 * 64, name='smol_weight_gen', use_bias=False)
+
+        flow = tf.transpose(inputs, perm=[0, 2, 3, 1])
+        flow = tf.reshape(flow, [-1, 64, tf.shape(inputs)[1]])
+        # add positional encoding for each square to the input
+        if self.arc_encoding:
+            self.POS_ENC = apm.make_pos_enc()
+            positional_encoding = tf.broadcast_to(tf.convert_to_tensor(self.POS_ENC, dtype=flow.dtype),
+                                                  [tf.shape(flow)[0], 64, tf.shape(self.POS_ENC)[2]])
+            flow = tf.concat([flow, positional_encoding], axis=2)
+
+        # square embedding
+        flow = tf.keras.layers.Dense(embedding_size, kernel_initializer='glorot_normal',
+                                     kernel_regularizer=self.l2reg, activation=self.DEFAULT_ACTIVATION,
+                                     name='embedding')(flow)
+
+        # !!! input gate
+        flow = ma_gating(flow, name='embedding')
+
+        for i in range(self.encoder_layers):
+            flow, attn_wts_l = self.encoder_layer(flow, embedding_size, self.encoder_d_model,
+                                                  self.encoder_heads, self.encoder_dff,
+                                                  name='encoder_{}'.format(i + 1))
+        return flow, attn_wts_l
```

Comment:
This `attn_wts_l` for each encoder block should be added into an array and returned. Like in line 1449 below.

Reply:
added them all to the same list. Any attention visualizers will have to account for that :P